File: backend/src/openrouter/openrouter.service.ts
import { Injectable, Logger } from '@nestjs/common'
import { ConfigService } from '@nestjs/config'
import { EventEmitter2 } from '@nestjs/event-emitter'

interface OpenRouterMessage {
  role: 'system' | 'user' | 'assistant'
  content: string
}

interface StreamChunkEvent {
  sessionId: string
  messageId: string
  chunk: string
  finishReason: string | null
}

@Injectable()
export class OpenRouterService {
  private readonly logger = new Logger(OpenRouterService.name)
  private readonly apiKey: string
  private readonly referer: string
  private readonly baseUrl = 'https://openrouter.ai/api/v1/chat/completions'

  private readonly activeControllers = new Map<string, AbortController>()

  constructor(
    private readonly config: ConfigService,
    private readonly eventEmitter: EventEmitter2,
  ) {
    this.apiKey = this.config.getOrThrow<string>('OPENROUTER_API_KEY')
    this.referer = this.config.get<string>('OPENROUTER_REFERER') ?? 'https://morganai.example.com'
  }

  /**
   * Streams a chat completion from OpenRouter.
   * @param opts.sessionId   WebSocket room id
   * @param opts.messages    Full message history INCLUDING the new user message and system prompt
   * @param opts.model       Model ID e.g. "anthropic/claude-3.5-sonnet"
   * @param opts.temperature Optional sampling temperature
   */
  async streamCompletion(opts: {
    sessionId: string
    messageId: string
    messages: OpenRouterMessage[]
    model: string
    temperature?: number
  }): Promise<void> {
    const controller = new AbortController()
    this.activeControllers.set(opts.sessionId, controller)

    try {
      const response = await fetch(this.baseUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.apiKey}`,
          'HTTP-Referer': this.referer,
          'X-Title': 'MorganAI',
        },
        body: JSON.stringify({
          model: opts.model,
          messages: opts.messages,
          stream: true,
          temperature: opts.temperature ?? 0.8,
        }),
        signal: controller.signal,
      })

      if (!response.ok || !response.body) {
        const text = await response.text()
        throw new Error(`OpenRouter error ${response.status}: ${text}`)
      }

      const reader = response.body.getReader()
      const decoder = new TextDecoder('utf-8')
      let buffer = ''
      let fullContent = ''

      while (true) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        buffer = lines.pop() ?? ''

        for (const line of lines) {
          if (!line.startsWith('data:')) continue
          const data = line.slice(5).trim()
          if (data === '[DONE]') {
            this.emitChunk({
              sessionId: opts.sessionId,
              messageId: opts.messageId,
              chunk: '',
              finishReason: 'stop',
            })
            continue
          }

          try {
            const parsed = JSON.parse(data)
            const delta = parsed.choices?.[0]?.delta?.content ?? ''
            const finish = parsed.choices?.[0]?.finish_reason ?? null
            if (delta || finish) {
              fullContent += delta
              this.emitChunk({
                sessionId: opts.sessionId,
                messageId: opts.messageId,
                chunk: delta,
                finishReason: finish,
              })
            }
          } catch {
            // ignore malformed SSE lines
          }
        }
      }

      this.logger.log(`Stream completed for session ${opts.sessionId} — ${fullContent.length} chars`)
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : 'Unknown error'
      this.logger.error(`OpenRouter stream failed: ${message}`)
      this.eventEmitter.emit('chat.error', {
        sessionId: opts.sessionId,
        messageId: opts.messageId,
        error: message,
      })
    } finally {
      this.activeControllers.delete(opts.sessionId)
    }
  }

  abortSession(sessionId: string): void {
    const ctrl = this.activeControllers.get(sessionId)
    if (ctrl) {
      ctrl.abort()
      this.activeControllers.delete(sessionId)
      this.logger.warn(`Aborted stream for session ${sessionId}`)
    }
  }

  private emitChunk(payload: StreamChunkEvent): void {
    this.eventEmitter.emit('chat.chunk', payload)
  }
}

File: backend/app/services/ai/openrouter_service.py
"""Асинхронный HTTP-клиент для OpenRouter API (LLM + Vision)."""

import base64
import logging
from typing import Any, Dict, List, Optional

import httpx

from app.core.config import settings

logger = logging.getLogger(__name__)


class OpenRouterService:
    """Реальный сервис для вызова OpenRouter API."""

    def __init__(self, api_key: Optional[str] = None, base_url: Optional[str] = None, default_model: Optional[str] = None):
        self.api_key = api_key or settings.OPENROUTER_API_KEY
        self.base_url = base_url or settings.OPENROUTER_BASE_URL
        self.default_model = default_model or settings.OPENROUTER_DEFAULT_MODEL
        self.headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "HTTP-Referer": "https://morgan-ai.app",
            "X-Title": "Morgan AI",
        }

    async def chat(
        self,
        messages: List[Dict[str, Any]],
        model: Optional[str] = None,
        temperature: float = 0.8,
        max_tokens: Optional[int] = None,
        **extra_kwargs,
    ) -> str:
        """Отправляет запрос чата к OpenRouter."""
        payload = {
            "model": model or self.default_model,
            "messages": messages,
            "temperature": temperature,
        }
        if max_tokens:
            payload["max_tokens"] = max_tokens
        payload.update(extra_kwargs)

        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.post(
                f"{self.base_url}/chat/completions",
                headers=self.headers,
                json=payload,
            )
            response.raise_for_status()
            data = response.json()

        try:
            content = data["choices"][0]["message"]["content"]
            logger.info(f"[OpenRouter] ответ получен, модель={payload['model']}")
            return content
        except (KeyError, IndexError) as exc:
            logger.error(f"[OpenRouter] неожиданный формат ответа: {data}")
            raise RuntimeError("Invalid OpenRouter response") from exc

    async def describe_image(
        self,
        image_bytes: bytes,
        user_text: str = "",
        mime_type: str = "image/jpeg",
        model: Optional[str] = None,
    ) -> str:
        """Отправляет текст + изображение (base64) в LLM и возвращает описание/ответ."""
        b64 = base64.b64encode(image_bytes).decode("utf-8")
        messages = [
            {
                "role": "user",
                "content": [
                    {"type": "text", "text": user_text or "Опиши, что на изображении."},
                    {"type": "image_url", "image_url": {"url": f"data:{mime_type};base64,{b64}"}},
                ],
            }
        ]
        return await self.chat(messages, model=model, temperature=0.5, max_tokens=1024)

    async def generate_image_prompt(self, user_message: str) -> str:
        """Генерирует английский промпт для генерации изображения."""
        system_msg = (
            "Ты — помощник по созданию промптов для генерации изображений. "
            "Переведи запрос пользователя на английский язык и сделай его детальным, атмосферным и визуальным."
        )
        messages = [
            {"role": "system", "content": system_msg},
            {"role": "user", "content": f"Создай промпт для картинки: {user_message}"},
        ]
        return await self.chat(messages, temperature=0.7, max_tokens=512)
